emulator/callbacks.go

```diff
@@ -0,0 +1,126 @@
+/*
+ * Flow Emulator
+ *
+ * Copyright Flow Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package emulator
+
+import (
+	_ "embed"
+	"encoding/binary"
+	"fmt"
+	"strings"
+
+	"github.com/onflow/cadence"
+	jsoncdc "github.com/onflow/cadence/encoding/json"
+	flowsdk "github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+//go:embed templates/executeCallbackTransaction.cdc
+var executeCallbackScript []byte
+
+//go:embed templates/processCallbackTransaction.cdc
+var processCallbackScript []byte
+
+func processCallbackTransaction(
+	serviceAddress flow.Address,
+	parentID flow.Identifier,
+) flow.TransactionBody {
+	script := replaceSchedulerAddress(processCallbackScript, serviceAddress)
+
+	return *flow.NewTransactionBody().
+		SetScript(script).
+		SetComputeLimit(defaultTransactionMaxGasLimit).
+		SetPayer(serviceAddress).
+		SetReferenceBlockID(parentID)
+}
+
+func executeCallbackTransactions(
+	scheduleEvent []flowsdk.Event,
+	serviceAddress flow.Address,
+	parentID flow.Identifier,
+) ([]flow.TransactionBody, error) {
+	var transactions []flow.TransactionBody
+	script := replaceSchedulerAddress(executeCallbackScript, serviceAddress)
+
+	for _, e := range scheduleEvent {
+		limit, id, err := parseSchedulerProcessedEvent(e, serviceAddress)
+		if err != nil {
+			return nil, err
+		}
+
+		tx := flow.NewTransactionBody().
+			SetScript(script).
+			AddArgument(id).
+			SetPayer(serviceAddress).
+			SetReferenceBlockID(parentID).
+			SetComputeLimit(limit)
+
+		transactions = append(transactions, *tx)
+	}
+
+	return transactions, nil
+}
+
+// parseSchedulerProcessedEvent parses flow event that is emitted during scheduler
+// marking the callback as scheduled.
+// Returns:
+// - execution effort
+// - ID of the event encoded as bytes
+// - error in case the event type is not correct
+func parseSchedulerProcessedEvent(event flowsdk.Event, serviceAddress flow.Address) (uint64, []byte, error) {
+	callbackScheduledEvent := fmt.Sprintf(
+		"A.%s.UnsafeCallbackScheduler.CallbackProcessed",
+		serviceAddress.String(),
+	)
```

Comment:
Same as above: Use the functions provided by Cadence to construct a correct type ID, don't use string formatting